subxt/src/client/online_client.rs

```diff
@@ -204,20 +222,35 @@ impl<T: Config> OnlineClientT<T> for OnlineClient<T> {
 pub struct ClientRuntimeUpdater<T: Config>(OnlineClient<T>);
 
 impl<T: Config> ClientRuntimeUpdater<T> {
-    fn is_runtime_version_different(&self, new: &RuntimeVersion) -> bool {
+    pub fn is_runtime_version_different(&self, new: &RuntimeVersion) -> bool {
         let curr = self.0.inner.read();
         &curr.runtime_version != new
     }
 
+    pub async fn do_update(
+        &self,
+        new_runtime_version: RuntimeVersion,
+    ) -> Result<(), Error> {
+        // Fetch new metadata.
+        let new_metadata = self.0.rpc.metadata().await?;
```

Comment:
I guess it's possible that the RuntimeVersion passed in becomes out of sync with the metadata that's obtained, since you can hold onto some RuntimeVersion for a while and then later on you can call `do_update` with that, and it goes and sets it without really checking anything, and fetches the _latest_ metadata to go along with it.

What would you think about fetching metadata in `perform_runtime_updates` so eg:

```rust
struct Update {
    version: RuntimeVersion,
    metadata: Metadata
}

// as now; the "easy" thing.
async fn perform_runtime_updates(&self) -> result<(), Error>;
// return a subscription which returns update payloads as updates are required.
async fn runtime_updates(&self) -> Result<Subscription<Update>, Error>;
// apply some Update payload.
fn apply_update(&self, update: Update) -> Result<(), Error>;
 ```

ie keep metadata and version bundled to try and ensure things will always be consistent?

Reply:
Example usage might then just be:

```rust
while let Some(Ok(update)) = updater.runtime_updates() {
    updater.apply_update(update)
}
```

(and I guess instead of exposing "is_update_different" you could filter the returned Subscription to do that for you perhaps? Hmm. Because if you've already obtained metadata then doing this check is a bit late. So I see why you obtained it separately.. hmmm)

Reply:
> It's only ever not different for the first result though I think

Yes, because on startup the runtime version is fetched by `subxt` and then when the subscription is created then the node will send out the initial value 

I think 3) is my cup of tea here your proposed API looks a bit nicer IMO

Reply:
> Yes, because on startup the runtime version is fetched by subxt and then when the subscription is created then the node will send out the initial value

Oh I mean as in, when you subscribe to runtime updates you always get the current version as the first message that comes back, so it's almost always equal :)

3) sounds good to me :)

Reply:
> Oh I mean as in, when you subscribe to runtime updates you always get the current version as the first message that comes back, so it's almost always equal :)

Yes, that was my point as subxt fetches the current runtime version on startup and once you subscribe to the runtime version then Substrate sends out the initial value of the runtime version on each subscription and those will be equal "all the time in practice" when a new subscription is created. After that I don't expect these numbers to be equal.

I ended up with 1) anyway I don't like having to use the `RwLock` in the `stream impl` let's be naive for now.

Reply:
Sounds good to me :)